src/Utils/PathSegment.vala

```diff
@@ -0,0 +1,302 @@
+/**
+ * Copyright (c) 2022 Alecaddd (https://alecaddd.com)
+ *
+ * This file is part of Akira.
+ *
+ * Akira is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+
+ * Akira is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+
+ * You should have received a copy of the GNU General Public License
+ * along with Akira. If not, see <https://www.gnu.org/licenses/>.
+ *
+ * Authored by: Ashish Shevale <shevaleashish@gmail.com>
+*/
+
+/*
+ * Following is a generic structure used for storing all types of segments
+ * is a path.
+ * type : refers to the type of segment.
+ * curve_begin : first point of a curve or end point of line.
+ * tangent_1, tangent_2 : control points for controlling curves.
+ * curve_end : last point of a curve
+ *
+ * We can create a different types of segments using this rep as
+ * 1. Symmetric Bezier Curve (both control points are equi-distant from curve_begin)
+ * 2. Asymmetric Bezier Curve (control points can be at different distances from curve_begin)
+ * 3. Quadratic Curve (single control point)
+ * 4. Line (pretty obvious)
+ */
+public struct Akira.Utils.PathSegment {
+    public Lib.Modes.PathEditMode.Type type;
+
+    // This point is used for drawing lines.
+    // Store point in segment_begin to save space, but use the alias to make code readable.
+    public Geometry.Point line_end {
+        get {
+            return curve_begin;
+        }
+        set {
+            curve_begin = value;
+        }
+    }
+
+    public Geometry.Point last_point {
+        get {
+            if (type == Lib.Modes.PathEditMode.Type.LINE) {
+                return line_end;
+            } else if (type == Lib.Modes.PathEditMode.Type.QUADRATIC) {
+                return curve_begin;
+            }
+
+            return curve_end;
+        }
+    }
+
+    // These points are used for drawing curves.
+    public Geometry.Point curve_begin;
+    public Geometry.Point tangent_1;
+    public Geometry.Point tangent_2;
+    public Geometry.Point curve_end;
+
+    public PathSegment () {
+        type = Lib.Modes.PathEditMode.Type.NONE;
```

Comment:
Why do you initialize these values, it seems very inefficient to me. You don't need the values to make sense until you use them.